### PR TITLE
i70–Create Hyperlinks in Slide Deck and BOM Columns

### PR DESCRIPTION
**Pull request by eshwaribhide**

### CODE REVIEW

**Comment by eshwaribhide** [comment]
one thing needs to be changed, need to add a check for if there is any data in slide deck/bom columns only then add the text, otherwise there are these empty links
will do soon

File: server-code/Projects.js [hunk 1]
```diff
@@ -2,12 +2,48 @@
 Document: JS code specific for projects
 */
 
+const slideDeckStr = "Slide Deck";
```
Comment (jamescd18) [inline server-code/Projects.js:5]: I love this! we haven't specifically called out constants in other areas. just two things:
1. let's use standard naming scheme for constants of something like `SLIDE_DECK_STR`
2. let's add a single line comment or something above to help explain what they are

File: server-code/Projects.js [hunk 2]
```diff
@@ -2,12 +2,48 @@
...
+
+/**
+ * Creates an HTML hyperlink, given the display text, that requests the given url.
+ * 
+ * @param {String} url – The url to link to
+ * @param {String} displayText – The text that will be displayed
+ * @return {String} – The corresponding HTML hyperlink
+ */
+function getHTMLLink(url, displayText) {
+    return displayText.link(url);
```
Comment (jamescd18) [inline server-code/Projects.js:27]: this is cool, I didn't know it existed. unfortunately, [Mozilla MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/link) seems to indicate that it's a deprecated feature, so we should probably just do it manually with a string template

**Review by jamescd18 — CHANGES_REQUESTED** [review]
Looking fantastic! just what we wanna go for. a few minor code changes

File: server-code/Projects.js (hunk 2 above)
Comment (kevinyu328) [reply server-code/Projects.js:27]: Agreed. Was going to point this out just now as well.

File: server-code/Projects.js (hunk 1 above)
Comment (eshwaribhide) [reply server-code/Projects.js]: Resolved by commit b8d322e26867e709212c3ada222fd6fc394bde42

File: server-code/Projects.js (hunk 2 above)
Comment (eshwaribhide) [reply server-code/Projects.js:30]: Resolved by commit 8ad9b33794c32a6aa1b117af787129ffee61e860

**Comment by eshwaribhide** [comment]
> one thing needs to be changed, need to add a check for if there is any data in slide deck/bom columns only then add the text, otherwise there are these empty links
> will do soon

Resolved by commit b8b31bbe7afde7600cacb509cdd0eb33dc47521c

File: server-code/Projects.js (hunk 2 above)
Comment (eshwaribhide) [reply server-code/Projects.js]: Actually, resolved by af6f9d4d4973afd66a5224d174052a511297b541

File: server-code/Projects.js [hunk 3]
```diff
@@ -2,12 +2,52 @@
...
+ * @return {void}
+ */
+function transformToHyperLinks(data) {
+    var headers = data[0];
+    var slideDeckColIdx = findIdx(SLIDE_DECK_STR, headers);
+    var bomColIdx = findIdx(BOM_STR, headers);
+    for (var rowIdx = 1; rowIdx < data.length; rowIdx++) {
+        slideDeckURL = data[rowIdx][slideDeckColIdx];
+        bomURL = data[rowIdx][bomColIdx];
+        if (slideDeckURL && bomURL) {
```
Comment (jamescd18) [inline server-code/Projects.js:46]: this is good! quick thought: what happens when bomURL is populated but slideDeckURL is an empty string? are they ever null values? I think the answer is no, but I could be wrong. Would it be better to do the if as two inline if statements sorta like:
 `slideDeckURL.length > 0 ? getHTMLLink(slideDeckURL, SLIDE_DECK_STR) : "" // converts url to HTML or leaves as empty string`

**Review by jamescd18 — CHANGES_REQUESTED** [review]
looking really good, I think it's almost there. quick question attached to the specific line

File: server-code/Projects.js (hunk 3 above)
Comment (kevinyu328) [reply server-code/Projects.js:46]: Agreed. This implementation would only work if both URLs are populated. Combining your method of checking for null/empty strings and James' inline statement, you could also use something like:
`slideDeckURL ? getHTMLLink(slideDeckURL, SLIDE_DECK_STR) : "" // converts url to HTML or leaves as empty string`

File: server-code/Projects.js [hunk 4]
```diff
@@ -2,12 +2,52 @@
...
+
+/**
+ * Creates an HTML hyperlink, given the display text, that requests the given url.
+ * 
+ * @param {String} url – The url to link to
+ * @param {String} displayText – The text that will be displayed
+ * @return {String} – The corresponding HTML hyperlink
+ */
+function getHTMLLink(url, displayText) {
+    var html = `<a href="` + url + `">` + displayText + `</a`
```
[the remark this answers is not in the document]
Comment (kevinyu328) [reply server-code/Projects.js:28]: Very minor; still works without the edit but I figure just to make things complete.

**Review by kevinyu328 — CHANGES_REQUESTED** [review]
Nice work! Just some small changes and I think we're good to go!

File: server-code/Projects.js (hunk 3 above)
Comment (eshwaribhide) [reply server-code/Projects.js:46]: Yeah I had just assumed that probably there would not be such a case but I guess it’s possible that may not always be the case 
Will change

**Review by kevinyu328 — APPROVED** [review]
Looks perfect to me!

**Review by jamescd18 — APPROVED** [review]

**Pull request merged**